utilities/ps.py

```diff
@@ -0,0 +1,99 @@
+#!/usr/bin/env python3
+from __future__ import annotations
+
+import subprocess
+from dataclasses import dataclass
+
+
+@dataclass
+class Service:
+    name: str
+    bindings: set[tuple[int, int]]
+
+    def print(self):
+        """
+        Print the formatted output for the service. The output contains the following:
+        - name (in bold)
+        - each URL (with the correct protocol) and the container port to which it maps
+        It specially handles the singular NGINX port 9443 which serves over ``https``.
+        """
+
+        print(f"\033[1m{self.name}:\033[0m")
+        for host_port, container_port in self.bindings:
+            proto = "http"
+            if self.name == "proxy" and container_port == 9443:
+                proto = "https"
+            print(f"- {proto:>5}://0.0.0.0:{host_port} (→ {container_port})")
+
+
+def get_ps() -> list[str]:
+    """
+    Invoke Docker Compose to get the configuration for all services. The config is
+    returned as a yaml string.
+
+    :return: the output printed by the subprocess to STDOUT
+    """
+
+    proc = subprocess.run(
+        [
+            "docker",
+            "ps",
+            "--filter",
+            "name=openverse",
+            "--format",
+            "{{.Names}}\n\t{{.Ports}}\n",
+        ],
+        check=True,
+        capture_output=True,
+        text=True,
+    )
+    return proc.stdout.split("\n")
+
+
+def parse_ps() -> list[Service]:
+    """
+    Convert the yaml output given by Docker Compose config into a list of services and
+    their port mappings.
+    :return: a list of running services with their port
+    """
+
+    services: list[Service] = []
+
+    lines = get_ps()
+    data_iter = iter([stripped for line in lines if (stripped := line.strip())])
+    # Iterate over the first two items in the list in pairs
+    # https://stackoverflow.com/a/48347320/3277713
+    for service, ports in zip(data_iter, data_iter):
+        # Services are usually of the form openverse-<name>-1 or openverse_<name>_1,
+        # but we can't always tell which one is used.
+        strip_char = service.removeprefix("openverse")[0]
+        name = service.split(strip_char, 1)[1].rsplit(strip_char, 1)[0]
```

Comment:
The underscore issue is happening because the catalog `image` is set to `openverse_catalog` rather than `openverse-catalog`. Everything else before adding the catalog to the monorepo uses the hypen, hence the discrepancy.

I think this code can be simplifed though :slightly_smiling_face: 

```suggestion
        name = service[len("openverse") + 1:].split("_")
```

Provided we always use `_` for the names of images aside from the openverse prefix, that should do it.

Ideally we'd update the catalog image name to be consistent with the other ones, though.

Reply:
I'm not sure this will work as-written, since we have some services (like `upstream_db`) which have an underscore in them as well. I think this might be using the image name rather than the container name though, at which point some further simplification might be possible. I'll try to see if I can simplify this.

Reply:
Okay, I tried using the docker image ID but that actually reduced the information available for some services (like `cache`, since the image ID is just `redis:<version>`. Also I think Docker does put different split characters depending on the version/machine - the example from the issue description (copied below, from my desktop) has underscores whereas all the other examples from folks (and from the computer I'm currently using) uses dashes:

```
$ j up
env COMPOSE_PROFILES="api,ingestion_server,frontend" docker-compose -f docker-compose.yml up -d
openverse_plausible_ch_1 is up-to-date
openverse_es_1 is up-to-date
openverse_upstream_db_1 is up-to-date
openverse_db_1 is up-to-date
openverse_cache_1 is up-to-date
openverse_plausible_db_1 is up-to-date
openverse_indexer_worker_1 is up-to-date
openverse_web_1 is up-to-date
openverse_plausible_1 is up-to-date
Starting openverse_proxy_1 ... 
Starting openverse_proxy_1 ... done

Port configuration for the services can be displayed by installing the Python package 'pyyaml'.
```

I'm going to try one more approach to potentially maximize the information we convey though.

Reply:
> Also I think Docker does put different split characters depending on the version/machine

That's just bizarre, TIL!

Reply:
I _know_ 😭